amaranth/lib/crc/__init__.py

```diff
@@ -0,0 +1,417 @@
+
+    For a reference of standard CRC parameter sets, refer to:
+
+    * `reveng`_'s catalogue, which uses an identical parameterisation,
+    * `crcmod`_'s list of predefined functions, but remove the leading '1'
+      from the polynominal, XOR the "Init-value" with "XOR-out" to obtain
+      ``initial_crc``, and where "Reversed" is True, set both ``reflect_input``
+      and ``reflect_output`` to True,
+    * `CRC Zoo`_, which contains only polynomials; use the "explicit +1"
+      form of polynomial but remove the leading '1'.
+
+    .. _reveng: https://reveng.sourceforge.io/crc-catalogue/all.htm
+    .. _crcmod: http://crcmod.sourceforge.net/crcmod.predefined.html
+    .. _CRC Zoo: https://users.ece.cmu.edu/~koopman/crc/
+
+    Many commonly used CRC algorithms are available in the
+    :py:mod:`~amaranth.lib.crc.catalog` module, which includes
+    all entries in the `reveng`_ catalogue.
+
+    To create a :py:class:`Parameters` instance, call the :py:class:`Algorithm`
+    object with the required data width, which defaults to 8 bits.
+
+    Parameters
+    ----------
+    crc_width : int
+        Bit width of CRC word. Also known as "width" in the Williams model.
+    polynomial : int
+        CRC polynomial to use, ``crc_width`` bits long, without the implicit
+        ``x**crc_width`` term. Polynomial is always specified with the highest
+        order terms in the most significant bit positions; use
+        ``reflect_input`` and ``reflect_output`` to perform a least
+        significant bit first computation.
+    initial_crc : int
+        Initial value of CRC register at reset. Most significant bit always
+        corresponds to the highest order term in the CRC register.
+    reflect_input : bool
+        If True, the input data words are bit-reflected, so that they are
+        processed least significant bit first.
+    reflect_output : bool
+        If True, the output CRC is bit-reflected, so the least-significant bit
+        of the output is the highest-order bit of the CRC register.
+        Note that this reflection is performed over the entire CRC register;
+        for transmission you may want to treat the output as a little-endian
+        multi-word value, so for example the reflected 16-bit output 0x4E4C
+        would be transmitted as the two octets 0x4C 0x4E, each transmitted
+        least significant bit first.
+    xor_output : int
+        The output CRC will be the CRC register XOR'd with this value, applied
+        after any output bit-reflection.
+    """
+    def __init__(self, *, crc_width, polynomial, initial_crc, reflect_input,
+                 reflect_output, xor_output):
+        self.crc_width = int(crc_width)
+        self.polynomial = int(polynomial)
+        self.initial_crc = int(initial_crc)
+        self.reflect_input = bool(reflect_input)
+        self.reflect_output = bool(reflect_output)
+        self.xor_output = int(xor_output)
+
+        assert self.crc_width > 0
```

Comment:
These should be runtime type/range checks, not assertions. Please follow the same structure for range checks as elsewhere in the codebase.